flask/app.py

```diff
@@ -1133,16 +1150,21 @@ def register_error_handler(self, code_or_exception, f):
 
     @setupmethod
     def _register_error_handler(self, key, code_or_exception, f):
-        if isinstance(code_or_exception, HTTPException):
-            code_or_exception = code_or_exception.code
-        if isinstance(code_or_exception, integer_types):
-            assert code_or_exception != 500 or key is None, \
-                'It is currently not possible to register a 500 internal ' \
-                'server error on a per-blueprint level.'
-            self.error_handler_spec.setdefault(key, {})[code_or_exception] = f
-        else:
-            self.error_handler_spec.setdefault(key, {}).setdefault(None, []) \
-                .append((code_or_exception, f))
+        """
+        :type key: None|str
+        :type code_or_exception: int|T<=Exception
+        :type f: callable
+        """
+        if isinstance(code_or_exception, HTTPException):  # old broken behavior
+            raise ValueError(
+                'Tried to register a handler for an exception instance {0!r}. '
+                'Handlers can only be registered for exception classes or HTTP error codes.'
+                .format(code_or_exception))
+        
+        exc_class, code = self._ensure_exc_class(code_or_exception)
+        
+        handlers = self.error_handler_spec.setdefault(key, {}).setdefault(code, {})
```

Comment:
i thought about that as well and thought that it’s clearer this way. all handlers are neatly grouped and accessing them in debug mode using `self.error_handler_spec[None][403]` will list you everything.

but i’m not particularly attached to this.

PS: we can preserve partial backwards compat by using a dict subclass for the `error_handler_spec[foobar]` level by redefining `__setitem__`:

```python
class BackwardsCompatDict(dict):
    def __init__(self, app):
        super().__init__()
        self.app = app
    def __setitem__(self, key, value):
        if callable(value) and not isinstance(value, dict):
            exc_class, code = self.app._ensure_exc_class(key)
            self.setdefault(code, {})[exc_class] = value
        else:
            super().__setitem__(key, handler_or_handlerdict)

app.error_handler_spec[None][403] = forbidden_handler
assert app.error_handler_spec[None][403][Forbidden] is forbidden_handler
```